src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1673,6 +1673,11 @@ AssignTasksToConnections(DistributedExecution *execution)
 				WorkerSession *session =
 					FindOrCreateWorkerSession(workerPool, connection);
 
+				if (PQstatus(connection->pgConn) == CONNECTION_BAD)
+				{
+					ConnectionStateMachine(session);
```

Comment:
I'm still not entirely comfortable with the placement of this function call, since there is a lot of error handling logic that relies on task assignments. We also have some (slightly clunky) clean-up logic for re-partitioning jobs in PG_CATCH now. What if we indiscriminately call ConnectionStateMachine for each session at the beginning of PG_TRY in RunDistributedExecution?

Reply:
Updated